fabric-registry-sync-v0/src/client/java/net/fabricmc/fabric/impl/client/registry/sync/RegistryRemovalChecker.java

```diff
@@ -0,0 +1,227 @@
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.mojang.datafixers.util.Pair;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import net.minecraft.block.Block;
+import net.minecraft.registry.Registries;
+import net.minecraft.registry.Registry;
+import net.minecraft.registry.RegistryKey;
+import net.minecraft.registry.RegistryKeys;
+import net.minecraft.state.property.Property;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.JsonHelper;
+import net.minecraft.util.Util;
+
+import net.fabricmc.fabric.api.event.registry.RegistryAttribute;
+import net.fabricmc.fabric.api.event.registry.RegistryAttributeHolder;
+
+public final class RegistryRemovalChecker {
+	private static final int VERSION = 1;
+	public static final Logger LOGGER = LoggerFactory.getLogger(RegistryRemovalChecker.class);
+	private static final Gson GSON = new Gson();
+	public static final String FILE_NAME = "fabric_registry_removal_check.json";
+	private static final boolean DISABLED = Boolean.getBoolean("fabric.registry.debug.disableRemovalCheck");
+	private final Set<String> missingNamespaces;
+	private final Set<RegistryKey<?>> missingKeys;
+	private final Set<Pair<Block, String>> missingBlockStates = new HashSet<>();
+
+	@SuppressWarnings("unchecked, rawtypes")
+	public RegistryRemovalChecker(JsonObject root) {
+		JsonObject json = root.getAsJsonObject("entries");
+		Set<RegistryKey<?>> keys = new HashSet<>();
+
+		for (Map.Entry<String, JsonElement> registry : json.entrySet()) {
+			Identifier registryId = Identifier.tryParse(registry.getKey());
+
+			if (registryId == null) continue;
+
+			RegistryKey<? extends Registry<?>> registryRef = RegistryKey.ofRegistry(registryId);
+
+			for (Map.Entry<String, JsonElement> namespacedEntries : registry.getValue().getAsJsonObject().entrySet()) {
+				for (JsonElement entry : namespacedEntries.getValue().getAsJsonArray()) {
+					Identifier entryId = Identifier.of(namespacedEntries.getKey(), entry.getAsString());
+
+					if (entryId == null) continue;
+
+					keys.add(RegistryKey.of((RegistryKey) registryRef, entryId));
+				}
+			}
+
+			Registry<?> clientRegistry = Registries.REGISTRIES.get(registryId);
+
+			if (clientRegistry != null) keys.removeAll(clientRegistry.getKeys());
+		}
+
+		for (Map.Entry<String, JsonElement> blockNsEntry : root.getAsJsonObject("blockStates").entrySet()) {
+			for (Map.Entry<String, JsonElement> blockEntry : blockNsEntry.getValue().getAsJsonObject().entrySet()) {
+				Identifier id = Identifier.of(blockNsEntry.getKey(), blockEntry.getKey());
+
+				if (id == null || !Registries.BLOCK.containsId(id)) continue;
```

Comment:
I know that the missing block will be caught above, but it might still be good to include it in the missing states.